pkg/apis/serving/v1alpha1/revision_lifecycle.go

```diff
@@ -270,3 +271,84 @@ func (r *Revision) GetLastPinned() (time.Time, error) {
 func (rs *RevisionStatus) duck() *duckv1beta1.Status {
 	return &rs.Status
 }
+
+// PropagateDeploymentStatus takes the Deployment status and applies its values
+// to the Revision status.
+func (rs *RevisionStatus) PropagateDeploymentStatus(original *appsv1.DeploymentStatus) {
+	ds := transformDeploymentStatus(original)
+	cond := ds.getCondition(deploymentConditionReady)
+	if cond == nil {
+		return
+	}
+
+	switch cond.Status {
+	case corev1.ConditionUnknown:
+		revCondSet.Manage(rs).MarkUnknown(RevisionConditionResourcesAvailable, cond.Reason, cond.Message)
+	case corev1.ConditionTrue:
+		revCondSet.Manage(rs).MarkTrue(RevisionConditionResourcesAvailable)
+	case corev1.ConditionFalse:
+		revCondSet.Manage(rs).MarkFalse(RevisionConditionResourcesAvailable, cond.Reason, cond.Message)
+	}
+}
+
+var depCondSet = apis.NewLivingConditionSet(
+	deploymentConditionProgressing,
+	deploymentConditionReplicaSetReady,
+)
+
+const (
+	// deploymentConditionReady means the underlying deployment is ready.
+	deploymentConditionReady = apis.ConditionReady
+	// deploymentConditionReplicaSetready inverts the underlying deployment's
+	// ReplicaSetFailure condition.
+	deploymentConditionReplicaSetReady apis.ConditionType = "ReplicaSetReady"
+	// deploymentConditionProgressing reflects the underlying deployment's
+	// Progressing condition.
+	deploymentConditionProgressing apis.ConditionType = "Progressing"
+)
+
+type deploymentStatus struct {
+	duckv1beta1.Status `json:",inline"`
+}
+
+func (ds *deploymentStatus) getCondition(t apis.ConditionType) *apis.Condition {
+	return depCondSet.Manage(ds).GetCondition(t)
+}
+
+func (ds *deploymentStatus) initializeConditions() {
+	depCondSet.Manage(ds).InitializeConditions()
+	// The absence of this condition means no failure has occured.
+	depCondSet.Manage(ds).MarkTrue(deploymentConditionReplicaSetReady)
+}
+
+// transformDeploymentStatus transforms the kubernetes DeploymentStatus into a
+// duckv1beta1.Status that uses ConditionSets to propagate failures and expose
+// a top-level happy state, per our condition conventions.
+func transformDeploymentStatus(ds *appsv1.DeploymentStatus) *deploymentStatus {
```

Comment:
I wonder if this is something that's better suited for `pkg/apis/serving` similar to our k8s validation stuff?

Reply:
Yeah I wasn't sure where to put this. There's a `pkg/deployment`, too...

We could also put it in `pkg/apis/apps/...` to mirror where it lives under k/k... we will probably want to do something similar for pods as part of https://github.com/knative/serving/issues/4557

Reply:
We have k8s stuff under `pkg/apis/serving` for field masking and validation.  That seems as good as any until we come up with another use case where it'd be awkward.